### ISSUE TITLE
Questions about zscore

### ISSUE THREAD

**Issue by ystephenai** [hazdzz/STGCN#10]
Hi,
I have to admit that your code inspires me a lot and I have definitely stared it.
However, could you explain about this piece of code below in main_road_traffic.py about zscore? 
I knew that the zscore normalized the features across the data size, so if the input data's dimension is like [batch, features], each feature will be normalized to mean =0, std = 1.
But in your code here,  if I am right, the dimension of variable "train" is [sequence length, num_of nodes]. If you apply zscore here, each node will be normalized to (0, 1) ? that's actually wierd for me because if we normalize each nodes accross the time step,won't we loss the spatial information? 

```python
train, val, test = dataloader.load_data(data_path, len_train, len_val)
    zscore = preprocessing.StandardScaler()
    train = zscore.fit_transform(train)
    val = zscore.transform(val)
    test = zscore.transform(test)
```

**Comment by hazdzz** [comment]
> Hi,
> I have to admit that your code inspires me a lot and I have definitely stared it.
> However, could you explain about this piece of code below in main_road_traffic.py about zscore?
> I knew that the zscore normalized the features across the data size, so if the input data's dimension is like [batch, features], each feature will be normalized to mean =0, std = 1.
> But in your code here, if I am right, the dimension of variable "train" is [sequence length, num_of nodes]. If you apply zscore here, each node will be normalized to (0, 1) ? that's actually wierd for me because if we normalize each nodes accross the time step,won't we loss the spatial information?
> 
> ```python
> train, val, test = dataloader.load_data(data_path, len_train, len_val)
>     zscore = preprocessing.StandardScaler()
>     train = zscore.fit_transform(train)
>     val = zscore.transform(val)
>     test = zscore.transform(test)
> ```

I don't understand what you mean. Why using z-score would lose the spatial information?

**Comment by hazdzz** [comment]
> Hi,
> I have to admit that your code inspires me a lot and I have definitely stared it.
> However, could you explain about this piece of code below in main_road_traffic.py about zscore?
> I knew that the zscore normalized the features across the data size, so if the input data's dimension is like [batch, features], each feature will be normalized to mean =0, std = 1.
> But in your code here, if I am right, the dimension of variable "train" is [sequence length, num_of nodes]. If you apply zscore here, each node will be normalized to (0, 1) ? that's actually wierd for me because if we normalize each nodes accross the time step,won't we loss the spatial information?
> 
> ```python
> train, val, test = dataloader.load_data(data_path, len_train, len_val)
>     zscore = preprocessing.StandardScaler()
>     train = zscore.fit_transform(train)
>     val = zscore.transform(val)
>     test = zscore.transform(test)
> ```

mean is not 0, and std is not 1. Why did you think mean is 0 and std is 1? Could you explain your opinions more specifically? You really confused me

**Comment by ystephenai** [comment]
Ok I will give you an example of what I am thinking:

For example, if your training matrix is:
[[1, 2, 3],
[2, 5, 3],
[0, 1, 5],
[2, 5, 4]]
after doing transformation **zscore.fit_transform(train)** you will get matrix **train**:
array([[-0.30151134, -0.70014004, -0.90453403],
       [ 0.90453403,  0.98019606, -0.90453403],
       [-1.50755672, -1.26025208,  1.50755672],
       [ 0.90453403,  0.98019606,  0.30151134]])

and you can find the sum of each columns equals to 0.
which in your case means that the mean of each position across time steps equals to 0.

**Comment by ystephenai** [comment]
And I don't really think that would be correct, for example, if one position is always (across time T) higher than another position, how do you capture that spatio information if you normalize the mean equals to 0. Could you get what I mean?

**Comment by hazdzz** [comment]
I think you're overthinking. For the z-score normalized matrix you offered as

[[-0.30151134, -0.70014004, -0.90453403],
[ 0.90453403, 0.98019606, -0.90453403],
[-1.50755672, -1.26025208, 1.50755672],
[ 0.90453403, 0.98019606, 0.30151134]]

the spatial information is not lost through time.

**Comment by hazdzz** [comment]
Btw, this is not my paper. If you think using z-score is wrong. Please tell the author. Here is the author's repository:
https://github.com/VeritasYin/STGCN_IJCAI-18

**Comment by ystephenai** [comment]
ok, but actually my example is not well-explained my idea.
if we consider a more extreme case:
[[0,1,2],
[0,1,2],
[0,1,2]]
after z-score or any normalization method, we always get 
[0,0,0]
[0,0,0]
[0,0,0]
and then we totally lose the spatial information in this case

**Comment by ystephenai** [comment]
But thanks you anyway for your explaination. That might be something unusual and I will test it on my own. 
Thanks you again.

**Comment by hazdzz** [comment]
> ok, but actually my example didn't well explain my thought.
> if we consider a more extreme case:
> [[0,1,2],
> [0,1,2],
> [0,1,2]]
> after z-score or any normalization method, we always get
> [0,0,0]
> [0,0,0]
> [0,0,0]
> and then we totally lose the spatial information in this case

No. The norm matrix as
[0,0,0]
[0,0,0]
[0,0,0]
is impossible. You misunderstand the formulation of normalization.

**Comment by ystephenai** [comment]
Wait, but why is it impossible? I just tried it with sklearn. it actually gives that result with all 0.

**Comment by ystephenai** [comment]
I don't know how to say, maybe you are right.

**Comment by hazdzz** [comment]
Finally, I knew what you mean.

import numpy as np
from scipy import stats

a = np.array([
    [0,1,2],
    [0,1,2],
    [0,1,2]
])

b = stats.zscore(a, axis=0)
print(b)

[[nan nan nan]
 [nan nan nan]
 [nan nan nan]]

In this case, you should use parameter axis = 1. As below

b = stats.zscore(a, axis=1)
print(b)

[[-1.22474487  0.          1.22474487]
 [-1.22474487  0.          1.22474487]
 [-1.22474487  0.          1.22474487]]

**Comment by hazdzz** [comment]
As I mentioned before, this extreme case would not occur in reality. So, you shall not worry about it.

**Comment by ystephenai** [comment]
> Finally, I knew what you mean.
> 
> import numpy as np
> from scipy import stats
> 
> a = np.array([
> [0,1,2],
> [0,1,2],
> [0,1,2]
> ])
> 
> b = stats.zscore(a, axis=0)
> print(b)
> 
> [[nan nan nan]
> [nan nan nan]
> [nan nan nan]]
> 
> In this case, you should use parameter axis = 1 as below
> 
> b = stats.zscore(a, axis=1)
> print(b)
> 
> [[-1.22474487 0. 1.22474487]
> [-1.22474487 0. 1.22474487]
> [-1.22474487 0. 1.22474487]]

yes, but don't you think you should use the same stats.zscore in your reconstructed code? indead of using zscore.fit_transform? that actually give two different results

**Comment by hazdzz** [comment]
No, as I mentioned, this extreme case would not occur in reality. So I don't need to change my code, and you don't worry about it.

**Comment by ystephenai** [comment]
ok
